Adds a check management command to make sure our merging is working

fredkingham opened the pull request:
Adds a managment command that makes sure...

* We do not have MRNs that are both in Demographics.hospital_number and MergedMRN.

* We do not have MRNs that are marked as active and merged upstream that do not have any merged MRNs

* We do not have MRNs that are marked as inactive upstream and active in our system.

Commit:
Adds a check management command to make sure our merging is working

Adds a managment command that makes sure...

* We do not have MRNs that are both in Demographics.hospital_number and
  MergedMRN.

* We do not have MRNs that are marked as active and merged upstream that
  do not have any merged MRNs

* We do not have MRNs that are marked as inactive upstream and active in
  our system.

davidmiller commented on intrahospital_api/management/commands/check_merged_patients.py:
```diff
@@ -0,0 +1,88 @@
+def check():
+    our_merged_mrns = set(elcid_models.MergedMRN.objects.values_list('mrn', flat=True))
+    our_demographics_mrns = set(elcid_models.Demographics.objects.all().values_list('hospital_number', flat=True))
+    our_merged_demographics_mrns = set(elcid_models.Demographics.objects.all().values_list(
```
Is this definitely different from the variable above? 
(I might be code blind)

davidmiller commented:
What's the context this is expected to execute & how do we know when it's found something?
Surely it should either raise exceptions _or_ send to logger.error not both?

If this is useful it's as something running automatically after the latest merge, so with a try, catch, email block...

In that context I'd add it to a cron job with an && (0 exit code) and remove 'thresholds'... 
(Unless there's a reason the threshold cases will always exist?)

Also, not sure the checks in the code match the list in the description here or the docstring? (can they all match)

Commit:
Simplifies the check merged patients command.

Simplifies check merge patient so we only check if there are any missing
active patients or any missing inactive patients

Commit:
Changes the check_merge command to explicitly ignore flawed rows

fredkingham commented:
These are all good points.

The reason for the threshold is that we know there are flawed upstream rows so we need to allow for these. I've changed it to explicitly ignore them based on id.

How this runs, I had envisioned it running once a week, just as a check sum to make sure everything works as expected but I think we should talk about this.

Commit:
Improves comments on chech_merged_patients

Commit:
Send a descriptive email rather than logger.error

Also don't raise a ValueError, handle all data issues by sending an
email.

Commit:
Handle flawed active MRNs in the check_merged_patients cmd

There are a lot of active MRNs with no merge comments, these can be
ignored for the purpose of this command.

There a alot of acitve MRNs where the upstream data is flawed. These
should be check on an individual basis for the time being. All current
flawed patients within elcid have been checked and they are an upstream
data issue and not ours.

fredkingham commented:
This mgmt command takes ~4 seconds to run

Commit:
Adds another active MRN to ignore in check_merged_patients

davidmiller commented on intrahospital_api/management/commands/check_merged_patients.py:
```diff
@@ -0,0 +1,232 @@
+"""
+Checks that our Merged MRN cron job is running.
```
That's not strictly what it does? 
that would look like `ps aux | grep merge_patients_since` ?

davidmiller commented on intrahospital_api/management/commands/check_merged_patients.py:
```diff
@@ -0,0 +1,232 @@
+        'hospital_number', flat=True
+    ))
+    intersection = our_merged_mrns.intersection(our_demographics_mrns)
+    if len(intersection) > 0:
```
Is this check listed in any of the docstrings? I can't see it

Commit:
Improves comments in check_merged_patients

Commit:
Adds 2 additional checks to check_merged_patients

Adds 2 additional checks to check_merged patients.

* Makes sure that every MRN with a merged MRN in our system is an MRN marked as an active merged mrn upstream.
* Makes sure that every merged MRN in our system is an MRN marked as inactive upstream

Commit:
Moves the check_merged_patient into monitoring

Moves the check_merged_patient mgmt command into monitoring.

davidmiller committed:
Update check_merged_patients.py

davidmiller commented on plugins/monitoring/management/commands/check_merged_patients.py:
```diff
@@ -0,0 +1,266 @@
+"""
+
+
+def send_email(title):
```
don't you want this to be utils now?

Commit:
Merge branch 'cerner-merge-branch' into chech-merge-patients

Commit:
Use utils.send_email in check_merged_patients

We now have a new utility for sending emails to admins so lets use this rather
than duplicate code.

Commit:
Merge branch 'chech-merge-patients' of github.com:openhealthcare/elcid-rfh into chech-merge-patients

Commit:
Merge branch 'cerner-merge-branch' into chech-merge-patients
